## Mint Automation
| Test | Result |
|:--- |:--- |
| mint-erasure.sh | :heavy_check_mark: |
| mint-compress-encrypt-dist-erasure.sh | :heavy_check_mark: |
| mint-pools.sh | :x: [more...](#ab4501eb-fc72-4cf8-977a-f727e5dcf66e) |
<a name="ab4501eb-fc72-4cf8-977a-f727e5dcf66e"></a>
### 16816-3dddb77/mint-pools.sh.log:
```
Running with
SERVER_ENDPOINT:      15.15.15.5:31500
ACCESS_KEY:           minio
SECRET_KEY:           ***REDACTED***
ENABLE_HTTPS:         0
SERVER_REGION:        us-east-1
MINT_DATA_DIR:        /mint/data
MINT_MODE:            full
ENABLE_VIRTUAL_STYLE: 0
RUN_ON_FAIL:          0

To get logs, run 'docker cp 4f3214bfb049:/mint/log /tmp/mint-logs'

(1/14) Running aws-sdk-go tests ... done in 8 seconds
(2/14) Running aws-sdk-java tests ... done in 2 seconds
(3/14) Running aws-sdk-php tests ... done in 43 seconds
(4/14) Running aws-sdk-ruby tests ... done in 9 seconds
(5/14) Running awscli tests ... done in 2 minutes and 24 seconds
(6/14) Running healthcheck tests ... done in 0 seconds
(7/14) Running mc tests ... done in 24 seconds
(8/14) Running minio-go tests ... done in 56 seconds
(9/14) Running minio-java tests ... done in 37 seconds
(10/14) Running minio-js tests ... done in 52 seconds
(11/14) Running minio-py tests ... FAILED in 2 minutes and 11 seconds
{
  "name": "minio-py:test_stat_object",
  "status": "FAIL",
  "args": {
    "bucket_name": "minio-py-test-c4cfe561-e285-4ce7-93f0-7967e7891e9d",
    "object_name": "2697ae28-952b-424f-b5e5-8eb5dfd2e0d4-metadata",
    "length": 11534336,
    "data": "LimitedRandomReader(11 * MB)",
    "metadata": {
      "X-Amz-Meta-Testing": "value"
    },
    "content_type": "application/octet-stream"
  },
  "message": "('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))",
  "error": "Traceback (most recent call last):\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 703, in urlopen\n    httplib_response = self._make_request(\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 449, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 444, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.8/http/client.py\", line 1348, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.8/http/client.py\", line 316, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.8/http/client.py\", line 277, in _read_status\n    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n  File \"/usr/lib/python3.8/socket.py\", line 669, in readinto\n    return self._sock.recv_into(b)\nConnectionResetError: [Errno 104] Connection reset by peer\n\nDuring handling of the above exception, another exception occurred:\n\nTraceback (most recent call last):\n  File \"/mint/run/core/minio-py/tests.py\", line 126, in _call_test\n    func(log_entry, *args, **kwargs)\n  File \"/mint/run/core/minio-py/tests.py\", line 820, in test_stat_object\n    _test_stat_object(log_entry, sse)\n  File \"/mint/run/core/minio-py/tests.py\", line 794, in _test_stat_object\n    result = _CLIENT.put_object(\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 1766, in put_object\n    raise exc\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 1725, in put_object\n    upload_id = self._create_multipart_upload(\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 1565, in _create_multipart_upload\n    response = self._execute(\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 403, in _execute\n    return self._url_open(\n  File \"/usr/local/lib/python3.8/dist-packages/minio/api.py\", line 266, in _url_open\n    response = self._http.urlopen(\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/poolmanager.py\", line 376, in urlopen\n    response = conn.urlopen(method, u.request_uri, **kw)\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 787, in urlopen\n    retries = retries.increment(\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/util/retry.py\", line 550, in increment\n    raise six.reraise(type(error), error, _stacktrace)\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/packages/six.py\", line 769, in reraise\n    raise value.with_traceback(tb)\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 703, in urlopen\n    httplib_response = self._make_request(\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 449, in _make_request\n    six.raise_from(e, None)\n  File \"<string>\", line 3, in raise_from\n  File \"/usr/local/lib/python3.8/dist-packages/urllib3/connectionpool.py\", line 444, in _make_request\n    httplib_response = conn.getresponse()\n  File \"/usr/lib/python3.8/http/client.py\", line 1348, in getresponse\n    response.begin()\n  File \"/usr/lib/python3.8/http/client.py\", line 316, in begin\n    version, status, reason = self._read_status()\n  File \"/usr/lib/python3.8/http/client.py\", line 277, in _read_status\n    line = str(self.fp.readline(_MAXLINE + 1), \"iso-8859-1\")\n  File \"/usr/lib/python3.8/socket.py\", line 669, in readinto\n    return self._sock.recv_into(b)\nurllib3.exceptions.ProtocolError: ('Connection aborted.', ConnectionResetError(104, 'Connection reset by peer'))\n",
  "duration": 113858
}
(11/14) Running s3cmd tests ... done in 18 seconds
(12/14) Running s3select tests ... done in 5 seconds
(13/14) Running versioning tests ... done in 3 minutes and 12 seconds

Executed 13 out of 14 tests successfully.
```
Deleting image on docker hub
Deleting image locally

--- a/docs/debugging/reorder-disks/go.mod
+++ b/docs/debugging/reorder-disks/go.mod
@@ -1,0 +1,5 @@
+module github.com/minio/minio/docs/debugging/reorder-disks
+
+go 1.19
+
+require github.com/minio/pkg v1.6.4

--- a/docs/debugging/reorder-disks/main.go
+++ b/docs/debugging/reorder-disks/main.go
@@ -1,0 +1,209 @@
+package main
+
+import (
+	"bufio"
+	"encoding/json"
+	"errors"
+	"flag"
+	"fmt"
+	"io/ioutil"
+	"log"
+	"net/url"
+	"os"
+	"path/filepath"
+	"strings"
+	"syscall"
+
+	"github.com/minio/pkg/ellipses"
+)
+
+type xl struct {
+	This string     `json:"this"`
+	Sets [][]string `json:"sets"`
+}
+
+type format struct {
+	ID string `json:"id"`
+	XL xl     `json:"xl"`
+}
+
+func getMountMap() (map[string]string, error) {
+	result := make(map[string]string)
+
+	mountInfo, err := os.Open("/proc/self/mountinfo")
+	if err != nil {
+		return nil, err
+	}
+	defer mountInfo.Close()
+
+	scanner := bufio.NewScanner(mountInfo)
+	for scanner.Scan() {
+		s := strings.Split(scanner.Text(), " ")
+		if len(s) != 11 {
+			return nil, errors.New("unsupport /proc/self/mountinfo format")
+		}
+		result[s[2]] = s[9]
+	}
+
+	if err := scanner.Err(); err != nil {
+		return nil, err
+	}
+
+	return result, nil
+}
+
+func getDiskUUIDMap() (map[string]string, error) {
+	result := make(map[string]string)
+	err := filepath.Walk("/dev/disk/by-uuid/",
+		func(path string, info os.FileInfo, err error) error {
+			if err != nil {
+				return err
+			}
+			realPath, err := filepath.EvalSymlinks(path)
+			if err != nil {
+				return err
+			}
+			result[realPath] = strings.TrimPrefix(path, "/dev/disk/by-uuid/")
+			return nil
+		})
+	if err != nil {
+		return nil, err
+	}
+	return result, nil
+}
+
+type localDisk struct {
+	index int
+	path  string
+}
+
+func getMajorMinor(path string) (string, error) {
+	var stat syscall.Stat_t
+	if err := syscall.Stat(path, &stat); err != nil {
+		return "", err
+	}
+
+	major := (stat.Dev & 0x00000000000fff00) >> 8
+	major |= (stat.Dev & 0xfffff00000000000) >> 32
+	minor := (stat.Dev & 0x00000000000000ff) >> 0
+	minor |= (stat.Dev & 0x00000ffffff00000) >> 12
+
+	return fmt.Sprintf("%d:%d", major, minor), nil
+}
+
+func filterLocalDisks(node, args string) ([]localDisk, error) {
+	var result []localDisk
+
+	argP, err := ellipses.FindEllipsesPatterns(args)
+	if err != nil {
+		return nil, err
+	}
+	exp := argP.Expand()
+
+	if node == "" {
+		for index, e := range exp {
+			result = append(result, localDisk{index: index, path: strings.Join(e, "")})
+		}
+	} else {
+		for index, e := range exp {
+			u, err := url.Parse(strings.Join(e, ""))
+			if err != nil {
+				return nil, err
+			}
+			if strings.Contains(u.Host, node) {
+				result = append(result, localDisk{index: index, path: u.Path})
+			}
+		}
+	}
+
+	return result, nil
+}
+
+func getFormatJSON(path string) (format, error) {
+	formatJSON, err := ioutil.ReadFile(filepath.Join(path, ".minio.sys/format.json"))
+	if err != nil {
+		return format{}, err
+	}
+	var f format
+	err = json.Unmarshal(formatJSON, &f)
+	if err != nil {
+		return format{}, err
+	}
+
+	return f, nil
+}
+
+func getDiskLocation(f format) (string, error) {
+	for i, set := range f.XL.Sets {
+		for j, disk := range set {
+			if disk == f.XL.This {
+				return fmt.Sprintf("%d-%d", i, j), nil
+			}
+		}
+	}
+	return "", errors.New("disk not found")
+}
+
+func main() {
+
+	var node, args string
+
+	flag.StringVar(&node, "filter-node", "", "Filter disks which belong to the given node")
+	flag.StringVar(&args, "args", "", "arguments passed to MinIO server")
+
+	flag.Parse()
+
+	localDisks, err := filterLocalDisks(node, args)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	if len(localDisks) == 0 {
+		log.Fatal("Fix --filter-node or/and --args to select local disks.")
+	}
+
+	format, err := getFormatJSON(localDisks[0].path)
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	setSize := len(format.XL.Sets[0])
+
+	expectedDisksName := make(map[string]string)
+	actualDisksName := make(map[string]string)
+
+	// Calculate the set/disk index
+	for _, disk := range localDisks {
+		expectedDisksName[fmt.Sprintf("%d-%d", disk.index/setSize, disk.index%setSize)] = disk.path
+		format, err := getFormatJSON(disk.path)
+		if err != nil {
+			log.Fatal(err)
+		}
+		foundDiskLoc, err := getDiskLocation(format)
+		if err != nil {
+			log.Fatal(err)
+		}
+		actualDisksName[foundDiskLoc] = disk.path
+	}
+
+	uuidMap, err := getDiskUUIDMap()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	mountMap, err := getMountMap()
+	if err != nil {
+		log.Fatal(err)
+	}
+
+	for loc, expectedDiskName := range expectedDisksName {
+		diskName := actualDisksName[loc]
+		mami, err := getMajorMinor(diskName)
+		if err != nil {
+			log.Fatal(err)
+		}
+		devName := mountMap[mami]
+		uuid := uuidMap[devName]
+		fmt.Printf("UID=%s\t%s\txfs\tdefaults,noatime\t0\t2\n", uuid, expectedDiskName)
+	}
+}

--- a/docs/debugging/reorder-disks/main.go
+++ b/docs/debugging/reorder-disks/main.go
@@ -1,3 +1,20 @@
+// Copyright (c) 2015-2023 MinIO, Inc.
+//
+// This file is part of MinIO Object Storage stack
+//
+// This program is free software: you can redistribute it and/or modify
+// it under the terms of the GNU Affero General Public License as published by
+// the Free Software Foundation, either version 3 of the License, or
+// (at your option) any later version.
+//
+// This program is distributed in the hope that it will be useful
+// but WITHOUT ANY WARRANTY; without even the implied warranty of
+// MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.  See the
+// GNU Affero General Public License for more details.
+//
+// You should have received a copy of the GNU Affero General Public License
+// along with this program.  If not, see <http://www.gnu.org/licenses/>.
+
 package main
 
 import (

